### PR TITLE
[AOTI][refactor] Refactor model runner API

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116114
* #116113
* __->__ #116047

Summary: 1) make proxy executor as a private member; 2) use std::string instead of char*

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D52301106](https://our.internmc.facebook.com/intern/diff/D52301106)

### CODE REVIEW

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52301106).

File: test/cpp/aot_inductor/test.cpp [hunk 1]
```diff
@@ -26,18 +26,18 @@ void test_aoti(const std::string& device) {
...
   if (device == "cuda") {
     runner = std::make_unique<torch::inductor::AOTIModelContainerRunnerCuda>(
-        model_so_path.c_str());
+        model_so_path);
   } else if (device == "cpu") {
     runner = std::make_unique<torch::inductor::AOTIModelContainerRunnerCpu>(
-        model_so_path.c_str());
+        model_so_path);
   } else {
     testing::AssertionFailure() << "unsupported device: " << device;
```
Comment (khabinov) [inline test/cpp/aot_inductor/test.cpp:42]: Feels like we should just port this test into `test_aot_inductor.py` eventually.

File: torch/csrc/inductor/aoti_runner/model_container_runner.cpp [hunk 2]
```diff
@@ -34,8 +34,22 @@ AOTIModelContainerRunner::AOTIModelContainerRunner(
...
+  #ifdef FBCODE_CAFFE2
+    std::string extern_kernel_nodes_path =
+        model_so_path.substr(0, model_so_path.size() - 3) + ".json";
+    if (fileExists(extern_kernel_nodes_path)) {
+      proxy_executor_handle_ =
+          std::make_unique<torch::aot_inductor::FbProxyExecutor>(
+              extern_kernel_nodes_path, is_cpu);
+  #endif
+  */
 }
```
Comment (khabinov) [inline torch/csrc/inductor/aoti_runner/model_container_runner.cpp:53]: Why the commented out code?

File: torch/csrc/inductor/aoti_runner/model_container_runner.h [hunk 3]
```diff
@@ -24,9 +24,8 @@ class TORCH_API AOTIModelContainerRunner {
   ~AOTIModelContainerRunner();
 
   std::vector<at::Tensor> run(
-      std::vector<at::Tensor> inputs,
-      AOTInductorStreamHandle cuda_stream_handle = nullptr,
-      AOTIProxyExecutorHandle proxy_executor_handle = nullptr);
+      std::vector<at::Tensor>& inputs,
+      AOTInductorStreamHandle cuda_stream_handle = nullptr);
```
Comment (khabinov) [inline torch/csrc/inductor/aoti_runner/model_container_runner.h:28]: `cuda_stream_handle` also seems to be irrelevant for cpu use cases.

File: torch/csrc/inductor/aoti_runner/model_container_runner.h [hunk 4]
```diff
@@ -58,6 +57,7 @@ class TORCH_API AOTIModelContainerRunner {
       swap_constant_buffer_func_{nullptr};
   decltype(&AOTInductorModelContainerGetCallSpec) get_call_spec_func_{nullptr};
   AOTInductorModelContainerHandle container_handle_ = nullptr;
+  AOTIProxyExecutorHandle proxy_executor_handle_ = nullptr;
```
Comment (khabinov) [inline torch/csrc/inductor/aoti_runner/model_container_runner.h:60]: Where do we set this variable?

File: torch/csrc/inductor/aoti_runner/model_container_runner.cpp (hunk 2 above)
Comment (desertfire) [reply torch/csrc/inductor/aoti_runner/model_container_runner.cpp:53]: I was trying to keep this as a reminder to add initialization for `proxy_executor_handle_` (partially answering your later question about it). Unfortunately, there is no equivalent of `FbProxyExecutor` in OSS yet. I will add a TODO at the declaration point of `proxy_executor_handle_` and remove the comments here.

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52301106).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52301106).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52301106).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52301106).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52301106).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52301106).

File: torch/csrc/inductor/aoti_runner/model_container_runner.h (hunk 4 above)
Comment (khabinov) [reply torch/csrc/inductor/aoti_runner/model_container_runner.h:64]: Thanks for adding the comment.

**Review by khabinov — APPROVED** [review]
